src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs

```
@@ -14,36 +15,114 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal
         {
             Input.Complete();
-            Output.Complete();
-            base.Dispose(disposing);
+            _output.Writer.Complete();
+
+            if (_outputTask != null)
+            {
+                // Wait for the output task to complete, this ensures that we've copied
+                // the application data to the underlying stream
+                await _outputTask;
+            }
         }
 
-        public override ValueTask DisposeAsync()
+        private async Task WriteOutputAsync()
         {
-            Input.Complete();
-            Output.Complete();
-            return base.DisposeAsync();
+            try
+            {
+                while (true)
+                {
+                    var result = await _output.Reader.ReadAsync();
+                    var buffer = result.Buffer;
+
+                    try
+                    {
+                        if (buffer.IsEmpty)
+                        {
+                            if (result.IsCompleted)
+                            {
+                                break;
+                            }
+
+                            await Stream.FlushAsync();
+                        }
+                        else if (buffer.IsSingleSegment)
+                        {
+                            await Stream.WriteAsync(buffer.First);
+                        }
+                        else
+                        {
+                            foreach (var memory in buffer)
+                            {
+                                await Stream.WriteAsync(memory);
+                            }
+                        }
+                    }
+                    finally
+                    {
+                        _output.Reader.AdvanceTo(buffer.End);
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                Log?.LogError(0, ex, $"{GetType().Name}.{nameof(WriteOutputAsync)}");
```

Given that both LibuvConnection and SocketConnection will not complete their PipeReaders with an error unless it was completely unexpected, it's tempting to make this a critical log. That's unless we're worried about something like SslStream throwing an Exception in the middle of WriteAsync despite there being no transport issue.

Maybe that is something we should be worried about... I'm just extra worried about swallowing errors now.

---

I'm going to leave it at error for now. I do agree that swallowing errors is a concern, but what else should we do. Debug.Assert? :trollface:

---

David already passed in a logger here. Critical logs are the new, much-improved Debug.Assert. Any critical log should cause the majority of our tests to fail.